storage/swift_storage.py

```diff
@@ -1,253 +1,254 @@
-                directories.append(container_object)
-            else:
-                files.append(container_object)
+    # cache get connections
+    def get_connection(self) -> swiftclient.client.Connection:
+        if not hasattr(self, "_connection"):
+            query = dict(parse_qsl(self._parsed_storage_uri.query))
 
-        return directories, files
+            auth_endpoint = query.get("auth_endpoint")
+            if auth_endpoint is None:
+                raise SwiftStorageError(f"Required field is missing: auth_endpoint")
 
-    def save_to_filename(self, file_path):
-        with open(file_path, "wb") as output_fp:
-            self.save_to_file(output_fp)
+            self.tenant_id = query.get("tenant_id")
+            if self.tenant_id is None:
+                raise SwiftStorageError(f"Required field is missing: tenant_id")
 
-    def save_to_file(self, out_file):
-        self._authenticate()
-        container_name, object_name = self._get_container_and_object_names()
+            region_name = query.get("region")
+            if region_name is None:
+                raise SwiftStorageError(f"Required field is missing: region_name")
 
-        for chunk in self._cloudfiles.fetch_object(
-                container_name, object_name, chunk_size=_LARGE_CHUNK):
-            out_file.write(chunk)
+            self.download_url_key = query.get("download_url_key")
 
-    def save_to_directory(self, destination_directory):
-        self._authenticate()
-        container_name, prefix = self._get_container_and_object_names()
+            os_options = {
+                "tenant_id": self.tenant_id,
+                "region_name": region_name
+            }
 
-        directories, files = self._list_container_objects(container_name, prefix)
+            auth, _ = self._parsed_storage_uri.netloc.split("@")
+            user, key = auth.split(":", 1)
 
-        for directory in directories:
-            if directory.name == prefix:
-                continue
+            if user == "":
+                raise SwiftStorageError(f"Missing username")
 
-            directory_path = directory.name.split('/', 1).pop()
-            target_directory = os.path.join(destination_directory, directory_path)
+            if key == "":
+                raise SwiftStorageError(f"Missing API key")
 
-            if not os.path.exists(target_directory):
-                os.makedirs(target_directory)
+            auth = v2.Password(
+                auth_url=auth_endpoint, username=user, password=key, tenant_name=self.tenant_id)
+
+            keystone_session = session.Session(auth=auth)
+
+            connection = swiftclient.client.Connection(
+                session=keystone_session, os_options=os_options, timeout=DEFAULT_SWIFT_TIMEOUT)
+            self._connection = connection
+        return self._connection
+
+    def get_container_and_object_names(self) -> Tuple[str, str]:
+        _, container = self._parsed_storage_uri.netloc.split("@")
```

Comment:
Again, can we just use `hostname` on `self._parsed_storage_uri`?

Reply:
@spiralman `hostname` returns the lowercased hostname. Are all container names case-insensitive?